File: ts/DivideAndConquer/SortAlgorithms/QuickSort/quickSort.ts

const list : number[] = [8,2,2,4,7,1,3,9,6,5,10,11,12,13,14,15]
const right : number = list.length -1;

function swap(list: number[], a:number, b:number): void{

    const temp : number = list[a];
    list[a] = list[b];
    list[b] = temp;
    
}

function partition(list : number[], left : number, right: number): number{

    const pivote: number = list[right];
    let orange: number = left - 1;

    for( let green = left; green <= right ; green++){

        if(list[green] <= pivote){

            orange++;
            if(green > orange){
                swap(list, green, orange);
            }
        }
    }

    return orange;

}

export function quickSort(list: number[], left: number, right:number): void{

    if(left < right){

        const pivote : number = partition(list, left, right)
        quickSort(list, left, pivote - 1)
        quickSort(list, pivote + 1 , right)
    }

}

console.log("La lista antes del sort=", list)

quickSort(list, 0 , right)

console.log("La lista despues del sort=", list)

File: c++/DivideAndConquer/SortAlgorithms/QuickSort/quickSort.cpp
// Video de referencia = https://www.youtube.com/watch?v=WprjBK0p6rw
#include <vector>
#include <iostream>
using namespace std;

vector<int> list = {8, 2, 4, 7, 1, 3, 9, 6, 5, 10, 11, 12, 13, 14, 15};
int n = list.size() - 1;

void swap(int &a, int &b)
{
    int temp = a;
    a = b;
    b = temp;
}

int partition(vector<int> &list, int left, int right)
{
    int pivot = list[right];
    int orange = left - 1;

    for (int green = left; green <= right; green++)
    {
        if (list[green] <= pivot)
        {
            orange++;
            if (green > orange)
            {
                swap(list[green], list[orange]);
            }
        }
    }
    return orange; // Regresamos la posicion donde quedo el pivote,
                   // se usa para dividir luego el arreglo
}

void quickSort(vector<int> &list, int left, int right)
{
    if (left < right)
    {
        int pivot = partition(list, left, right); // Aqui queda el pivote, lo usamos para dividir lo demas
        quickSort(list, left, pivot - 1);
        quickSort(list, pivot + 1, right);
    }
}

int main()
{

    cout << "Lista todavia no ordenada con el nesquick" << endl;
    for (int x : list)
        cout << x << " ";

    quickSort(list, 0, n);
    cout << "\nLa supuesta lista ordenada con quicksort=";
    for (int x : list)
        cout << x << " ";

    return 0;
}
